src/distributed_copy.c

```diff
@@ -0,0 +1,857 @@
+	List *convert_select;       /* list of column names (can be NIL) */
+	bool *convert_select_flags;         /* per-column CSV/TEXT CS flags */
+
+	/* these are just for error messages, see CopyFromErrorCallback */
+	const char *cur_relname;    /* table name for error messages */
+	int cur_lineno;             /* line number for error messages */
+	const char *cur_attname;    /* current att for error messages */
+	const char *cur_attval;     /* current att value for error messages */
+
+	/*
+	 * Working state for COPY TO/FROM
+	 */
+	MemoryContext copycontext;  /* per-copy execution context */
+
+	/*
+	 * Working state for COPY TO
+	 */
+	FmgrInfo *out_functions;    /* lookup info for output functions */
+	MemoryContext rowcontext;   /* per-row evaluation context */
+
+	/*
+	 * Working state for COPY FROM
+	 */
+	AttrNumber num_defaults;
+	bool file_has_oids;
+	FmgrInfo oid_in_function;
+	Oid oid_typioparam;
+	FmgrInfo *in_functions;     /* array of input functions for each attrs */
+	Oid *typioparams;           /* array of element types for in_functions */
+	int *defmap;                /* array of default att numbers */
+	ExprState **defexprs;       /* array of default att expressions */
+	bool volatile_defexprs;             /* is any of defexprs volatile? */
+	List *range_table;
+
+	/*
+	 * These variables are used to reduce overhead in textual COPY FROM.
+	 *
+	 * attribute_buf holds the separated, de-escaped text for each field of
+	 * the current line.  The CopyReadAttributes functions return arrays of
+	 * pointers into this buffer.  We avoid palloc/pfree overhead by re-using
+	 * the buffer on each cycle.
+	 */
+	StringInfoData attribute_buf;
+
+	/* field raw data pointers found by COPY FROM */
+
+	int max_fields;
+	char **raw_fields;
+
+	/*
+	 * Similarly, line_buf holds the whole input line being processed. The
+	 * input cycle is first to read the whole line into line_buf, convert it
+	 * to server encoding there, and then extract the individual attribute
+	 * fields into attribute_buf.  line_buf is preserved unmodified so that we
+	 * can display it in error messages if appropriate.
+	 */
+	StringInfoData line_buf;
+	bool line_buf_converted;            /* converted to server encoding? */
+	bool line_buf_valid;        /* contains the row being processed? */
+
+	/*
+	 * Finally, raw_buf holds raw data read from the data source (file or
+	 * client connection).	CopyReadLine parses this data sufficiently to
+	 * locate line boundaries, then transfers the data to line_buf and
+	 * converts it.	 Note: we guarantee that there is a \0 at
+	 * raw_buf[raw_buf_len].
+	 */
+#define RAW_BUF_SIZE 65536      /* we palloc RAW_BUF_SIZE+1 bytes */
+	char *raw_buf;
+	int raw_buf_index;          /* next byte to process */
+	int raw_buf_len;            /* total # of bytes stored */
+} CopyStateData;
+
+/* TODOL should be moved to copy.c */
+static StringInfo
+CopyGetLineBuf(CopyStateData *cs)
+{
+	return &cs->line_buf;
+}
+
+
+#define MAX_SHARDS 1001
```

Comment:
This should probably be named more appropriately, maybe INITIAL_CONNECTION_CACHE_SIZE and moved to the top below the #includes. You could also consider using MaxConnections instead.